addons/stock_account/models/product.py

```diff
@@ -50,210 +92,399 @@ def get_product_accounts(self, fiscal_pos=None):
+                }
+                candidate.write(candidate_vals)
+
+                qty_to_take_on_candidates -= qty_taken_on_candidate
+                tmp_value += value_taken_on_candidate
+                if float_is_zero(qty_to_take_on_candidates, precision_rounding=self.uom_id.rounding):
+                    break
+
+            # Get the estimated value we will correct.
+            remaining_value_before_vacuum = svl_to_vacuum.unit_cost * qty_taken_on_candidates
+            new_remaining_qty = svl_to_vacuum.remaining_qty + qty_taken_on_candidates
+            corrected_value = remaining_value_before_vacuum - tmp_value
+            svl_to_vacuum.write({
+                'remaining_qty': new_remaining_qty,
+            })
+
+            # Don't create a layer or an accounting entry if the corrected value is zero.
+            if svl_to_vacuum.currency_id.is_zero(corrected_value):
+                continue
+
+            corrected_value = svl_to_vacuum.currency_id.round(corrected_value)
+            move = svl_to_vacuum.stock_move_id
+            vals = {
+                'product_id': self.id,
+                'value': corrected_value,
+                'unit_cost': 0,
+                'quantity': 0,
+                'remaining_qty': 0,
+                'description': 'vacuum',
+                'stock_move_id': move.id,
+                'company_id': move.company_id.id,
+                'description': 'Revaluation of %s (negative inventory)' % move.picking_id.name or move.name,
+            }
+            vacuum_svl = self.env['stock.valuation.layer'].create(vals)
+
+            # Create the account move.
+            if self.valuation != 'real_time':
+                continue
+            vacuum_svl.stock_move_id._account_entry_move(
+                vacuum_svl.quantity, vacuum_svl.description, vacuum_svl.id, vacuum_svl.value
+            )
+
+    @api.model
+    def _svl_empty_stock(self, description, product_category=None, product_template=None):
+        impacted_product_ids = []
+        impacted_products = self.env['product.product']
+        products_orig_quantity_svl = {}
+
+        # get the impacted products
+        domain = [('type', '=', 'product')]
+        if product_category is not None:
+            domain += [('categ_id', '=', product_category.id)]
+        elif product_template is not None:
+            domain += [('product_tmpl_id', '=', product_template.id)]
         else:
-            action['res_model'] = 'stock.move'
-            action['domain'] = [('id', 'in', self.with_context(to_date=to_date).stock_fifo_manual_move_ids.ids)]
-            tree_view_ref = self.env.ref('stock_account.view_move_tree_valuation_at_date')
-            form_view_ref = self.env.ref('stock.view_move_form')
-            action['views'] = [(tree_view_ref.id, 'tree'), (form_view_ref.id, 'form')]
-        return action
+            raise ValueError()
+        products = self.env['product.product'].search_read(domain, ['quantity_svl'])
```

Comment:
why search_read if your browse right after ?

Reply:
search_read to keep the quantity_svl before the creation of the out layer
browse for comfort i guess :) i'm able to use _prepare_out_svl_vals easily